localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp

```diff
@@ -439,6 +439,18 @@ bool NDTScanMatcher::callback_sensor_points_main(
     add_regularization_pose(sensor_ros_time);
   }
 
+  // Warn if the lidar has gone out of the map range
```

Comment:
Can you add a reason why it is checked periodically in map_update_module but also here? 
My understanding is that this part is specifically checking for being out of the map during the waiting time for a service reply.

Reply:
@KYabuuchi In the current setting, it is checked once every 1 second. During that 1 second, if the vehicle travels too fast there is a chance that the lidar may get out of the map range (although in the current setting, it is very hard to occur, unless the vehicle speed is about 50 m/s ~ 180 km/h). So this check here is for such cases.

But I think if we reduce the period of the timer callback it would be better (like instead of 1 second we set it to 50 ms). However, doing so would raise another issue: while the map and NDT are being updated (which often takes more than 50ms), there is a chance that there is another should_update_map return true, thus triggers another update_map. Such duplicated operations are not necessary at all.

Reply:
@anhnv3991 Oh, I thought you implemented option 1 because option 2 wouldn't alert us if the vehicle goes out of the map while waiting for the map_loader service (which often takes more than 50ms).

I want to clarify what issues this PR is intended to address. I believe it tackles the following two issues:
1. The lack of checks during the waiting time for the map_loader service in MapUpdateModule.
2. The risk of the vehicle going out of the map within 1 second when dynamic_map_loading parameters are set to hard values (although this is rare with default parameters).

Please let me know if I misunderstood.

Your option 1 addresses both issues, so I think your option 1 is the natural choice. (However, there's also the option to check in the initial_pose_callback instead of the sensor_points_callback.)

Reply:
@KYabuuchi Actually, initially I thought this PR is to address No. 2 (which is pointed out by Ando-san). But No.1 is also a good point of this PR. Shortly, this PR warns us if the vehicle goes out of the map range anytime.